File: code/algo/getoffer/go/2/newest_timewindow_count.go
package main

type RecentCounter []int

func NewRecentCounter() (_ RecentCounter) { return }

func (q *RecentCounter) Ping(t int) int {
	*q = append(*q, t)
	for (*q)[0] < t-3000 {
		*q = (*q)[1:]
	}
	return len(*q)
}

File: code/algo/getoffer/c++/2/newest_timewindow_count.cpp
#include <deque>

class RecentCounter {
public:
  RecentCounter() {}

  int ping(int t) {
    q.push_back(t);
    while (q.front() < t - 3000) {
      q.pop_front();
    }
    return q.size();
  }

private:
  std::deque<int> q;
};
